### ISSUE TITLE
Stacks should have optional setting for required levels of Appsody/Docker/K8s

### ISSUE THREAD

**Issue by nastacio** [appsody/stacks#423]
**Is your feature request related to a problem? Please describe.**
Some adopters do not update their Docker and Kubernetes environments as frequently as appsody developers or operations teams.

Whenever appsody commands hit these unsupported versions, they fail with a sequence of messages about internal findings from these downstream dependencies, such as the Docker daemon chastising the end-user for some failed command inside a Dockerfile that is not visible to the end user, or the kube cluster saying some API resource inside a CRD does not exist, again, in a file that is not visible to the user.

**Describe the solution you'd like**
Appsody should allow stack developers to indicate the range of Appsody/Docker/Kubernetes they support, with hardcoded defaults matching the minimum accepted level by Appsody innards (e.g. appsody operator).

**Describe alternatives you've considered**
One could argue "documentation", but that is transferring a relative simple check in the code to yet-another manual check to be performed by all application developers on an ongoing basis.

**Additional context**
None.

**Comment by akihikokuroda** [comment]
This is the console output when I got the error.
```
[root@dyads1 project3]# appsody init nodejs-express
Running appsody init...
Downloading nodejs-express template project from https://github.com/appsody/stacks/releases/download/nodejs-express-v0.2.7/incubator.nodejs-express.v0.2.7.templates.simple.tar.gz
Download complete. Extracting files from nodejs-express.tar.gz
Setting up the development environment
Pulling docker image appsody/nodejs-express:0.2
Running command: docker pull appsody/nodejs-express:0.2
Trying to pull repository registry.access.redhat.com/appsody/nodejs-express ... 
Pulling repository registry.access.redhat.com/appsody/nodejs-express
Trying to pull repository docker.io/appsody/nodejs-express ... 
0.2: Pulling from docker.io/appsody/nodejs-express
Digest: sha256:36a60fa077a5e38d5275665f666c00f93e64896434d84efe8ebeab87b8c51210
Status: Image is up to date for docker.io/appsody/nodejs-express:0.2
[Warning] The stack image does not contain APPSODY_PROJECT_DIR. Using /project
Running command: docker run --rm --entrypoint /bin/bash appsody/nodejs-express:0.2 -c find /project -type f -name .appsody-init.sh
Successfully initialized Appsody project
[root@dyads1 project3]# appsody build
Extracting project from development environment
Pulling docker image appsody/nodejs-express:0.2
Running command: docker pull appsody/nodejs-express:0.2
Trying to pull repository registry.access.redhat.com/appsody/nodejs-express ... 
Pulling repository registry.access.redhat.com/appsody/nodejs-express
Trying to pull repository docker.io/appsody/nodejs-express ... 
0.2: Pulling from docker.io/appsody/nodejs-express
Digest: sha256:36a60fa077a5e38d5275665f666c00f93e64896434d84efe8ebeab87b8c51210
Status: Image is up to date for docker.io/appsody/nodejs-express:0.2
[Warning] The stack image does not contain APPSODY_PROJECT_DIR. Using /project
Running command: docker create --name project3-extract -v /root/project3/:/project/user-app appsody/nodejs-express:0.2
Running command: docker cp project3-extract:/project /root/.appsody/extract/project3
Running command: docker rm project3-extract -f
Project extracted to /root/.appsody/extract/project3
Running docker command: docker build -t project3 -f /root/.appsody/extract/project3/Dockerfile /root/.appsody/extract/project3
[Docker] Sending build context to Docker daemon 160.8 kB
[Docker] Step 1/21 : FROM node:10
[Docker]  ---> 636ef87129d6
[Docker] Step 2/21 : RUN apt-get update  && apt-get dist-upgrade -y  && apt-get clean  && echo 'Finished installing dependencies'
[Docker]  ---> Using cache
[Docker]  ---> 0c4c32ddc51f
[Docker] Step 3/21 : WORKDIR /project
[Docker]  ---> Using cache
[Docker]  ---> 07997cd424f4
[Docker] Step 4/21 : COPY ./package*.json ./
[Docker]  ---> Using cache
[Docker]  ---> fb445514f728
[Docker] Step 5/21 : RUN npm install --production
[Docker]  ---> Using cache
[Docker]  ---> 8641b597a0a6
[Docker] Step 6/21 : WORKDIR /project/user-app
[Docker]  ---> Using cache
[Docker]  ---> d8fa3d72fbbf
[Docker] Step 7/21 : COPY ./user-app/package*.json ./
[Docker]  ---> Using cache
[Docker]  ---> 94565aa0a009
[Docker] Step 8/21 : RUN npm install --production
[Docker]  ---> Using cache
[Docker]  ---> 9dd63f09ef89
[Docker] Step 9/21 : WORKDIR /project/user-app/node_modules
[Docker]  ---> Using cache
[Docker]  ---> cf0769b59f54
[Docker] Step 10/21 : FROM node:10-slim
[Docker]  ---> 8d33f30db9b5
[Docker] Step 11/21 : RUN apt-get update  && apt-get dist-upgrade -y  && apt-get clean  && echo 'Finished installing dependencies'
[Docker]  ---> Using cache
[Docker]  ---> 3591346d1a63
[Docker] Step 12/21 : COPY --chown=node:node . /project
[Docker] Unknown flag: chown
[Error] exit status 1

```

Here is the docer version output:
```
[root@dyads1 project3]# docker version
Client:
 Version:         1.13.1
 API version:     1.26
 Package version: docker-1.13.1-103.git7f2769b.el7.x86_64
 Go version:      go1.10.8
 Git commit:      7f2769b/1.13.1
 Built:           Fri Aug  2 10:19:53 2019
 OS/Arch:         linux/amd64

Server:
 Version:         1.13.1
 API version:     1.26 (minimum version 1.12)
 Package version: docker-1.13.1-103.git7f2769b.el7.x86_64
 Go version:      go1.10.8
 Git commit:      7f2769b/1.13.1
 Built:           Fri Aug  2 10:19:53 2019
 OS/Arch:         linux/amd64
 Experimental:    false

```

Here is the OS info:
```
[root@dyads1 project3]# cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 7.7 (Maipo)

```

**Comment by BarDweller** [comment]
I'd like to add appsody version to the list of things a stack can set required levels for  =) 

If a stack depends upon features only introduced at a given level of appsody, it would be nice for the stack to be able to prevent its own usage on earlier levels to prevent errors/loss of data etc.

**Comment by nastacio** [comment]
@BarDweller Agreed. I had it in the text, but missed it from the title. Added it.

**Comment by neeraj-laad** [comment]
I agree. We knew this was coming but never got to it. :)

**Comment by neeraj-laad** [comment]
We would add this information in `stack.yaml` / `index.yaml`.

- appsody
- docker
- kubernetes
- operator?

The cli could then read this information before specific commands are run with that stack and display a meaningful error message to the user.

**Comment by kylegc** [comment]
I agree this would be good to have. As described here, the stacks themselves may depend on specific versions. But the Appsody CLI itself should also check for minimum Docker and Kube versions. The challenge is determining the minimums. We can’t possibly test/support all the versions back to XYZ, or expect the stack creators to do this, but at least this gives us a way to adjust the min as we find issues.

What about buildah version?

When we say “kubernetes version”, do we mean the client `kubectl` or the kubernetes server? Or perhaps we need both.
The kubernetes version check should only be a WARNING because you can still use appsody run/debug/test/extract without kubernetes.

**Comment by Kamran64** [comment]
After discussions on slack/here, I think it would be good if:
- We add a new field to the `stack.yaml` titled `stack-requirements`. The stack provider can either supply a minimum version or version range which could be structured like so:
```
stack-requirements:
    appsody-version: >=0.3.0
    docker-version: >=12.1.0
    kubectl-version: >=1.9.0
    buildah-version: 1.5.0-2.3.0
```
- We could enforce the following technologies (subject to change or improvement): Appsody, Docker, Kubectl, k8s server, Buildah, Appsody Operator.

- We shouldn't be checking all of these at the same time. I.e. There's no need to block a user from using a stack if their Buildah version is out of date when they are only using Docker. Therefore:
     - CLI, Docker/Buildah: Init phase
     - k8s, Operator: Deploy

- I agree @kylegc about determining minimums. I'll find a suitable minimum as a baseline and then for stacks that might e.g. use `chown` I'll bump the versions.

**Comment by BarDweller** [comment]
Is there any way for a stack to contribute implementation of a check of it's own requirement ? Eg, say a java stack wanted to say 'if maven is installed locally, it must be at least version 3.3' .. or `java must be 1.11 for this stack' etc.. I'm sure other runtimes would have their own tests they'd want to perform.

**Comment by nastacio** [comment]
@BarDweller , IMO, stacks should not have dependencies on stack-specific tools installed to the developer's local environment. Personally, I like appsody that much better than CloudFoundry for not making me worry about trying to mimic the internals of the buildpack, errr...stack, in my local system.

**Comment by BarDweller** [comment]
While I agree stacks need to be functional without local tooling.. in practice, that's not how Java devs work, they are going to have stuff locally that they want to use.. (Be it VSCode/Eclipse etc) the way the Java ecosystem is setup, Java devs are going to have a JDK locally, are going to have Maven (or gradle, or sbt) locally, and most likely some sort of IDE. 

They will expect to be able to edit their appsody projects using that tooling, and that's where we'll need to be able to at least give a heads up to say 'hey, you're trying to use this appsody stack with JDK5, that's not such a great idea, we really need you to use JDK8 or above'. 

On the other hand, if they really have NO local tooling (they are probably actually a build engine ;)) and we can support all the appsody operations fine using the capabilities inside the container. 

Although it's feasible today to manipulate a Java appsody project using just vim, it's a horrific experience because no one really writes code in a non syntax aware editor these days, and without having the tooling locally (at least something that can understand Java syntax AND how to grok the dependency references from the maven pom), there's no way for a Java developer to have the little things they depend upon, like code completion, or type browsing. 

So I'm fine for a stack to say 'IF you have java, it MUST be this or above' rather than 'you MUST have java at this or above' .. but not saying anything at all is just going to lead to user pain when they try using a project with locally installed older versions of the toolset used by the stack itself.

**Comment by nastacio** [comment]
I did not argue a developer will not have tools. 

My point is that *appsody* should not depend on those tools for its own purposes.

Appsody printing out a manifest of tools used in the stack may be welcome as a reference point for developers building out their development environment, but that is a separate issue from the original request: letting the developer know that appsody dependencies are not met instead of letting downstream calls blow up with unintelligible messages.

**Comment by BarDweller** [comment]
So the Java stacks do not depend upon Java/Maven being available locally, but IF they are, they will be used to install the parent pom, to enable non-appsody local builds (eg IDEs, local maven usage) to correctly understand the dependencies of stack projects. 

Note: the dependency is not quite that clear cut, the stacks use maven wrapper to do the parent pom install during the init phase, but maven wrapper requires Java, which has to be present on the users system. If Java is not present, then maven is not used, and the parent pom will not be installed.

`appsody run` (along with `debug` `build` etc)  will function regardless of if the parent pom was installed to the users system. But this doesn't really address any serious development, which will be using an IDE to edit code, which will require the information from the parent pom. 

I still think there needs to be a way for a stack to be able to state this kind of stuff for Java, because building the project locally (outside of `appsody run`, using an IDE, or plain old `mvn package`) WILL be impacted by version issues such as Java Level etc. Otherwise Java devs will run into issues when stacks require JDK 11 with modules, and all they have is JDK8 etc. 

I would welcome a warning from appsody saying my Java or Maven was older than the stack expected. If nothing else it would be a heads up that I need to tidy up my environment. 

I think this is still in keeping with the original request, preventing a developer from failing to be able to develop using appsody due to downstream dependencies not meeting the expectations of the stack developers. 

Just because the errors occur when the user is using their own tooling rather than appsody should not be an issue here. Otherwise we are greatly lessening the value of appsody by suggesting that the only important issues are those triggered during execution of an appsody command, which would abandon the idea that appsody projects are seamlessly able to be built using regular normal developer tooling. (Eg, for Java, maven from a command line).

**Comment by nastacio** [comment]
I did not realize the Java stacks depended on the local availability of Java and Maven to the command-line from where `appsody init` is being called.

That said, there is still a conceptual divide between checks that can only be performed by appsody itself and the checks that are stack-specific and can only be performed by the stack code (I am assuming we don't want the appsody code being the one checking on maven, java, npm, etc).

Given the frequency of users hitting problems with obsolete versions of docker/k8s/appsody out there, having appsody checking for things that only appsody can "see" could be delivered first, and for now, stack code can continue checking dependency versions on their own.

We could discuss the IDE facet in that would-be new issue, since what is available for a CLI (via PATH or other environment variables is not guaranteed to be used by the IDE (e.g. Eclipse uses its own configuration to point at java and maven) .

**Comment by Kamran64** [comment]
I've raised https://github.com/appsody/appsody/pull/535 as a proof of concept to see how implementing this would work. Currently it only enforces versions of Docker and the Appsody CLI but it should be relatively simple to add more.

It's also been raised as a draft because I plan on cleaning it up a fair bit.

**Comment by Kamran64** [comment]
Closing this issue. This feature was implemented in 0.5.0, more details here: appsody/appsody#535
